include/hipSYCL/sycl/libkernel/cuda/group_functions.hpp

```diff
@@ -0,0 +1,240 @@
+__device__
+T shuffle_impl(T x, int id) {
+  return __shfl_sync(0xFFFFFFFF, x, id);
+}
+
+template<>
+__device__
+char shuffle_impl(char x, int id) {
+  return static_cast<char>(shuffle_impl(static_cast<int>(x), id));
+}
+
+template<typename T, int N>
+__device__
+sycl::vec<T, N> shuffle_impl(sycl::vec<T, N> x, int id) {
+  sycl::vec<T, N> ret{};
+
+  if constexpr (1 <= N)
+    ret.s0() = shuffle_impl(x.s0(), id);
+  if constexpr (2 <= N)
+    ret.s1() = shuffle_impl(x.s1(), id);
+  if constexpr (3 <= N)
+    ret.s2() = shuffle_impl(x.s2(), id);
+  if constexpr (4 <= N)
+    ret.s3() = shuffle_impl(x.s3(), id);
+  if constexpr (8 <= N) {
+    ret.s4() = shuffle_impl(x.s4(), id);
+    ret.s5() = shuffle_impl(x.s5(), id);
+    ret.s6() = shuffle_impl(x.s6(), id);
+    ret.s7() = shuffle_impl(x.s7(), id);
+  }
+  if constexpr (16 <= N) {
+    ret.s8() = shuffle_impl(x.s8(), id);
+    ret.s9() = shuffle_impl(x.s9(), id);
+    ret.sA() = shuffle_impl(x.sA(), id);
+    ret.sB() = shuffle_impl(x.sB(), id);
+    ret.sC() = shuffle_impl(x.sC(), id);
+    ret.sD() = shuffle_impl(x.sD(), id);
+    ret.sE() = shuffle_impl(x.sE(), id);
+    ret.sF() = shuffle_impl(x.sF(), id);
+  }
+
+  return ret;
+}
+
+} // namespace detail
+
+// broadcast
+template<typename T>
+HIPSYCL_KERNEL_TARGET
+T group_broadcast(sub_group g, T x,
+                  typename sub_group::linear_id_type local_linear_id = 0) {
+  return detail::shuffle_impl(x, local_linear_id);
+}
+
+template<typename T, int N>
+HIPSYCL_KERNEL_TARGET
+sycl::vec<T, N> group_broadcast(sub_group g, sycl::vec<T, N> x,
+                                typename sub_group::linear_id_type local_linear_id = 0) {
+  return detail::shuffle_impl(x, local_linear_id);
+}
+
+
+// barrier
+template<typename Group>
+HIPSYCL_KERNEL_TARGET
+inline void group_barrier(Group g, memory_scope fence_scope = Group::fence_scope) {
+  if (fence_scope == memory_scope::device) {
+    __threadfence_system();
+  }
+  __syncthreads();
+}
+
+template<>
+HIPSYCL_KERNEL_TARGET
+inline void group_barrier(sub_group g, memory_scope fence_scope) {
+  if (fence_scope == memory_scope::device) {
+    __threadfence_system();
+  } else if (fence_scope == memory_scope::work_group) {
+    __threadfence_block();
+  }
+  __syncwarp(); // not necessarily needed, but might improve performance
+}
+
+
+// any_of
+template<>
+HIPSYCL_KERNEL_TARGET
+inline bool group_any_of(sub_group g, bool pred) {
+  return __any_sync(0xFFFFFFFF, pred);
+}
+
+
+// all_of
+template<>
+HIPSYCL_KERNEL_TARGET
+inline bool group_all_of(sub_group g, bool pred) {
+  return __all_sync(0xFFFFFFFF, pred);
+}
+
+
+// none_of
+template<>
+HIPSYCL_KERNEL_TARGET
+inline bool group_none_of(sub_group g, bool pred) {
+  return !__any_sync(0xFFFFFFFF, pred);
+}
+
+
+// reduce
+template<typename T, typename BinaryOperation>
+HIPSYCL_KERNEL_TARGET
+T group_reduce(sub_group g, T x, BinaryOperation binary_op) {
```

Comment:
I could try to merge them into hiplike, I'll take a look. I originally split them up, because the shuffle-instruction is different. Now we have `shuffle_impl`, so I could give this a try.

Reply:
I'll keep them separated for now, since it is likely I will separate them again when optimizing. I'll keep it in mind though in case the optimized code can be implemented in a hiplike fashion :)